import_events.py

```diff
@@ -0,0 +1,236 @@
+            'epoch'::TIMESTAMP + timestamp * '1 second'::INTERVAL AS ts,
+            STRTOL(SUBSTRING({id_column} FROM 0 FOR 8), 16) % 100 AS cohort
+        FROM {temp_table}
+    )
+    WHERE cohort <= {percent}
+    AND ts::DATE = '{day}'::DATE
+    AND ts::DATE >= '{max_day}'::DATE - '{months} months'::INTERVAL;
+""".format(perm_table=TABLE_NAMES["perm"],
+           temp_table=TABLE_NAMES["temp"],
+           columns="{columns}",
+           id_column="{id_column}",
+           percent="{percent}",
+           day="{day}",
+           max_day="{max_day}",
+           months="{months}")
+
+Q_GET_TIMESTAMP = """
+    SELECT {which}(timestamp) FROM {table};
+""".format(which="{which}", table=TABLE_NAMES["temp"])
+
+Q_DELETE_EVENTS = """
+    DELETE FROM {table}
+    WHERE timestamp::DATE < '{day}'::DATE - '{months} months'::INTERVAL;
+""".format(table=TABLE_NAMES["perm"], day="{day}", months="{months}")
+
+Q_VACUUM_TABLES = """
+    END;
+    VACUUM FULL {table};
+    ANALYZE {table};
+""".format(table=TABLE_NAMES["perm"])
+
+def nop(*args):
+    pass
+
+def run(s3_prefix, event_type, temp_schema, temp_columns, perm_schema, perm_columns, id_column="uid",
+        before_import=nop, after_day=nop, after_import=nop, day_from=None, day_until=None):
+
+    def drop_temporary_table():
+        db.run(Q_DROP_TEMPORARY_TABLE.format(event_type=event_type))
+
+    def create_events_tables():
+        for rate in SAMPLE_RATES:
+            db.run(Q_CREATE_EVENTS_TABLE.format(event_type=event_type,
+                                                suffix=rate["suffix"],
+                                                schema=perm_schema))
+
+    def get_max_day():
+        result = db.one(Q_GET_MAX_DAY.format(event_type=event_type))
+        if result:
+            return datetime.strftime(result, "%Y-%m-%d")
+        return result
+
+    def is_candidate_day(day):
+        return (not day_from or day_from <= day) and (not day_until or day_until >= day)
+
+    def is_day_populated(day):
+        return bool(db.one(Q_CHECK_FOR_DAY.format(event_type=event_type, day=day)))
+
+    def get_unpopulated_days():
+        days = []
+        message = "FINDING UNPOPULATED DAYS"
+        if day_from:
+            message += " FROM {day_from}".format(day_from=day_from)
+        if day_until:
+            message += " UNTIL {day_until}".format(day_until=day_until)
+        print message
+        for key in s3.list(prefix=s3_prefix):
+            filename = path.basename(key.name)
+            day = "-".join(filename[:-4].split("-")[1:])
+            if is_candidate_day(day) and not is_day_populated(day):
+                days.append(day)
+        return days
+
+    def get_timestamp(which):
+        return db.one(Q_GET_TIMESTAMP.format(which=which, event_type=event_type))
+
+    def print_timestamp(which):
+        print "  {which} timestamp".format(which=which), get_timestamp(which)
+
+    def import_day(day):
+        print day
+        print "  COPYING CSV"
```

Comment:
I've been doing a *lot* of debugging of these scripts over the past couple of weeks and found the tweaks to the log messages helpful when scanning back through the output to see what's happened. Essentially I aimed for:

* Ouputting the minimal amount of information per line, to make it easy to parse.
* Indenting lines to emphasise hierarchy and thus reduce any need for repetition.